### ISSUE TITLE
pTest problem with __del__ and exceptions

### ISSUE THREAD

**Issue by KrisThielemans** [SyneRBI/SIRF#842]
I've currently got some failures (on Windows, Python 3.8), which gives me the following confusing output
```
....
.EException ignored in: <function pTest.__del__ at 0x0000015A35299B80>
Traceback (most recent call last):
  File "C:\Users\krisf\Documents\devel\buildVC\SIRF-SuperBuild-python\INSTALL\python\sirf\Utilities.py", line 267, in __del__
    raise ValueError(msg)
ValueError: 1 failures
...
======================================================================
ERROR: tests_four.test_main
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\krisf\Documents\devel\buildVC\SIRF-SuperBuild-python\INSTALL\python\sirf\Utilities.py", line 287, in check
    check_tolerance(expected, value, abs_tol, rel_tol)
  File "C:\Users\krisf\Documents\devel\buildVC\SIRF-SuperBuild-python\INSTALL\python\sirf\Utilities.py", line 240, in check_tolerance
    raise ValueError("|%.4g - %.4g| > %.3g" %
ValueError: |2.13e-08 - -0.001098| > 4.26e-11

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\krisf\miniconda3\envs\sirf\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\Users\krisf\Documents\devel\buildVC\SIRF-SuperBuild-python\sources\SIRF\src\xSTIR\pSTIR\tests\tests_four.py", line 42, in test_main
    test.check(1 - math.sqrt(acq_data.dot(acq_data)) / acq_data.norm())
  File "C:\Users\krisf\Documents\devel\buildVC\SIRF-SuperBuild-python\INSTALL\python\sirf\Utilities.py", line 292, in check
    raise ValueError(msg)
ValueError: +++ test 7 failed:|2.13e-08 - -0.001098| > 4.26e-11
```
I think the underlying reason (certainly for the first warning, possibly not for the double exception) is that `pTest.__del__` should not raise an exception, but exceptions in `__del__` are printed but ignored. In fact, `pTest.__del__` probably shouldn't exist at all as it isn't guaranteed to be called apparently.  See for instance https://stackoverflow.com/questions/1481488/what-is-the-del-method-how-to-call-it

In fact, should `runner` not test for exceptions?
https://github.com/SyneRBI/SIRF/blob/4bce83f0c2049011bc506584894c4be94061b190/src/common/Utilities.py#L340

**Comment by KrisThielemans** [comment]
@casperdcl maybe you could help?

**Comment by casperdcl** [comment]
Is the problem duplicate error messages? It doesn't make sense to me to re-raise in `__del__`.

I think the real problem is reinventing the wheel with our own test class. The idea seems to have been to collect failures and raise all of them in one go at the end (rather than failing early at the first exception).

Using another test suite to manage this (e.g. `pytest`) would make life much easier - can completely remove all this `pTest` stuff and just stick to simple `assert`s everywhere.

**Comment by KrisThielemans** [comment]
> Is the problem duplicate error messages? It doesn't make sense to me to re-raise in `__del__`.

There are possibly 2 independent problems, but I'm not sure. I now think that the "double exception" is because 
https://github.com/SyneRBI/SIRF/blob/554da3201e6bd5ca0fc4f3d606dd5d2ac9d56532/src/common/Utilities.py#L286-L292
but am not sure.

> I think the real problem is reinventing the wheel with our own test class.

I agree that `pytest` is nice and that having our own test class is possibly complicating things for the new developer, but it does have the convenient feature that it can record known values, which can be then be used for regression tests. (We are often using it for tests with known values as well though, which can cause problems, see #843, fixed in #844).

Of course, rewriting all tests is not something that we should embark on unless we have a good reason. 

I thought you had written the `pTest` stuff though. If not, let's just wait for @evgueni-ovtchinnikov's input (after the holiday!)
